hotdeploy causing problem when recompiling

### Description of the bug

When vaadin.frontend.hotdeploy is set to `true` and you recompile in a project (tested in the standard template provided by the Vaadin IntelliJ Plugin with gradle) the project will fail to recover from the recompilation and will spam the console with the same error over and over again.

### Expected behavior

The Project should recover normally or tell me if I have an incompatible feature enabled

### Minimal reproducible example

https://github.com/aless2003/bugTest2
Just open it with IntelliJ and recompile during the first start. It's a bit hard to reproduce unfortunately, just try it a couple of times (with a vaadinClean step in-between) and it should appear at least once. It seems to be a lot easier to reproduce right after Vite has been started, but in my day-to-day project it appears basically whenever the recompilation starts the "vaadinPrepareFrontend" task. If I find a better method to reproduce the bug, I'll let you know.

You should see something like this in the console
```
2024-08-06T01:42:24.723+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : [TypeScript] Found 0 errors. Watching for file changes.
2024-08-06T01:42:25.367+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : [vite] build/vaadin-dev-server-settings.json changed, restarting server...
2024-08-06T01:42:25.430+02:00 ERROR 10124 --- [nio-8080-exec-5] c.vaadin.flow.server.StaticFileServer    : Unable to load /VAADIN/@fs/V:/Tests/bugTest2/node_modules/.vite/deps/react-router-dom.js from the frontend dev server: Connection reset
2024-08-06T01:42:25.430+02:00 ERROR 10124 --- [io-8080-exec-10] c.vaadin.flow.server.StaticFileServer    : Unable to load /VAADIN/@fs/V:/Tests/bugTest2/node_modules/.vite/deps/construct-style-sheets-polyfill.js from the frontend dev server: Connection reset
2024-08-06T01:42:25.430+02:00 ERROR 10124 --- [nio-8080-exec-9] c.vaadin.flow.server.StaticFileServer    : Unable to load /VAADIN/@fs/V:/Tests/bugTest2/node_modules/.vite/deps/react-dom_client.js from the frontend dev server: Connection reset
2024-08-06T01:42:25.430+02:00 ERROR 10124 --- [nio-8080-exec-6] c.vaadin.flow.server.StaticFileServer    : Unable to load /VAADIN/@fs/V:/Tests/bugTest2/node_modules/@vaadin/bundles/vaadin.js from the frontend dev server: Connection reset
2024-08-06T01:42:25.430+02:00 ERROR 10124 --- [nio-8080-exec-7] c.vaadin.flow.server.StaticFileServer    : Unable to load /VAADIN/@fs/V:/Tests/bugTest2/node_modules/.vite/deps/react.js from the frontend dev server: Connection reset
2024-08-06T01:42:25.430+02:00 ERROR 10124 --- [nio-8080-exec-4] c.vaadin.flow.server.StaticFileServer    : Unable to load /VAADIN/@fs/V:/Tests/bugTest2/node_modules/.vite/deps/react_jsx-dev-runtime.js from the frontend dev server: Connection reset
2024-08-06T01:42:25.888+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : 
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :  ERROR(TypeScript)  Cannot find module 'Frontend/generated/flow/generated-flow-imports.js' or its corresponding type declarations.
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :  FILE  V:/Tests/bugTest2/src/main/frontend/generated/flow/Flow.tsx:28:27
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : 
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     26 |
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     27 | const flow = new _Flow({
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :   > 28 |     imports: () => import("Frontend/generated/flow/generated-flow-imports.js")
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :        |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     29 | });
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     30 |
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     31 | const router = {
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : 
2024-08-06T01:42:25.889+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : [TypeScript] Found 1 error. Watching for file changes.
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : 
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :  ERROR(TypeScript)  Cannot find module 'Frontend/generated/flow/generated-flow-imports.js' or its corresponding type declarations.
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :  FILE  V:/Tests/bugTest2/src/main/frontend/generated/flow/Flow.tsx:28:27
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : 
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     26 |
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     27 | const flow = new _Flow({
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :   > 28 |     imports: () => import("Frontend/generated/flow/generated-flow-imports.js")
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :        |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-06T01:42:27.157+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     29 | });
2024-08-06T01:42:27.158+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     30 |
2024-08-06T01:42:27.158+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   :     31 | const router = {
2024-08-06T01:42:27.158+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : 
2024-08-06T01:42:27.158+02:00  INFO 10124 --- [v-server-output] c.v.b.devserver.DevServerOutputTracker   : [TypeScript] Found 1 error. Watching for file changes.
```

### Versions

Hilla: 24.4.4
Flow: 24.4.4
Vaadin: 24.4.7
Copilot: 24.4.8
Copilot IDE Plugin: false
Java: Amazon.com Inc. 17.0.12
Java Hotswap: false
Frontend Hotswap: Enabled, using Vite
OS: amd64 Windows 11 10.0
Browser: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 Edg/128.0.0.0

--- a/flow-plugins/flow-gradle-plugin/src/functionalTest/kotlin/com/vaadin/gradle/VaadinSmokeTest.kt
+++ b/flow-plugins/flow-gradle-plugin/src/functionalTest/kotlin/com/vaadin/gradle/VaadinSmokeTest.kt
@@ -15,20 +15,22 @@
  */
 package com.vaadin.gradle
 
-import com.fasterxml.jackson.databind.JsonNode
-import com.vaadin.flow.internal.JacksonUtils
 import java.io.File
+import java.nio.file.Files
+import java.nio.file.StandardCopyOption
+import kotlin.io.path.div
+import kotlin.io.path.writeText
 import kotlin.test.assertContains
 import kotlin.test.expect
+import com.vaadin.flow.internal.JacksonUtils
 import com.vaadin.flow.internal.StringUtil
 import com.vaadin.flow.server.InitParameters
 import com.vaadin.flow.server.frontend.FrontendUtils
+import com.fasterxml.jackson.databind.JsonNode
 import org.gradle.testkit.runner.BuildResult
 import org.gradle.testkit.runner.TaskOutcome
 import org.junit.Before
 import org.junit.Test
-import java.nio.file.Files
-import java.nio.file.StandardCopyOption
 
 
 /**
@@ -567,6 +569,114 @@ class VaadinSmokeTest : AbstractGradleTest() {
         val result2 = testProject.build("--configuration-cache", "vaadinPrepareFrontend", "-Dvaadin.eagerServerLoad=true", checkTasksSuccessful = false)
         result2.expectTaskOutcome("vaadinPrepareFrontend", TaskOutcome.SUCCESS)
         assertContains(result.output, "Calculating task graph as no cached configuration is available for tasks: vaadinPrepareFrontend")
+    }
+
+    // When Hilla is detected, frontend hot deploy should be automatically
+    // enabled and as a consequence prepare frontend task should create react
+    // related files
+    @Test
+    fun vaadinPrepareFrontend_hillaAvailable_frontendHotDeployEnabled() {
+        // Setup a project local maven repo to simulate the presence of Hilla
+        val fakeHillaVersion = "0.0.0.localtest";
+        val projectLocalMavenRepo = testProject.newFolder("libs").absoluteFile
+        val hillaEndpointFolder =
+            projectLocalMavenRepo.toPath() / "com" / "vaadin" / "hilla-endpoint" / fakeHillaVersion
+        Files.createDirectories(hillaEndpointFolder)
+
+        // hilla-endpoint-stub.jar contains only stub for com.vaadin.hilla.EndpointController.class
+        val hillaEndpointJar =
+            hillaEndpointFolder / "hilla-endpoint-${fakeHillaVersion}.jar"
+        Files.copy(
+            File(javaClass.classLoader.getResource("hilla-endpoint-stub.jar").path).toPath(),
+            hillaEndpointJar, StandardCopyOption.REPLACE_EXISTING
+        )
+        val hillaEndpointPom =
+            hillaEndpointFolder / "hilla-endpoint-${fakeHillaVersion}.pom"
+        hillaEndpointPom.writeText(
+            """
+            <?xml version="1.0" encoding="UTF-8"?>
+            <project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd"
+                xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+                <modelVersion>4.0.0</modelVersion>
+                <groupId>com.vaadin</groupId>
+                <artifactId>hilla-endpoint</artifactId>
+                <version>${fakeHillaVersion}</version>
+                <name>Fake Hilla Endpoint</name>
+            </project>
+            """.trimIndent()
+        )
+        enableHilla()
+
+        testProject.buildFile.writeText(
+            """
+            plugins {
+                id 'war'
+                id 'com.vaadin'
+            }
+            repositories {
+                maven {
+                    url = '${projectLocalMavenRepo.toURI().toURL()}'
+                }
+                mavenLocal()
+                mavenCentral()
+                maven { url = 'https://maven.vaadin.com/vaadin-prereleases' }
+                flatDir {
+                   dirs("libs")
+                }
+            }
+            dependencies {
+                implementation("com.vaadin:flow:$flowVersion")
+                implementation("com.vaadin:hilla-endpoint:${fakeHillaVersion}")
+                providedCompile("jakarta.servlet:jakarta.servlet-api:6.0.0")
+                implementation("org.slf4j:slf4j-simple:$slf4jVersion")
+            }
+            vaadin {
+            }
+        """.trimIndent()
+        )
+
+        val flowTsx = File(
+            testProject.dir,
+            FrontendUtils.DEFAULT_PROJECT_FRONTEND_GENERATED_DIR + "flow/Flow.tsx"
+        )
+        val vaadinReactTsx = File(
+            testProject.dir,
+            FrontendUtils.DEFAULT_PROJECT_FRONTEND_GENERATED_DIR + "vaadin-react.tsx"
+        )
+
+        testProject.build("vaadinPrepareFrontend")
+        expect(
+            true,
+            "Expected Flow.tsx to be created when Hilla is available"
+        ) { flowTsx.exists() }
+        expect(
+            true,
+            "Expected vaadin-react.tsx to be created when Hilla is available"
+        ) { vaadinReactTsx.exists() }
+    }
+
+    // When Hilla is not available, frontend hot deploy should not be enabled
+    // by default and react related files should not be created
+    @Test
+    fun vaadinPrepareFrontend_hillaNotAvailable_frontendHotDeployNotEnabled() {
+        val flowTsx = File(
+            testProject.dir,
+            FrontendUtils.DEFAULT_PROJECT_FRONTEND_GENERATED_DIR + "flow/Flow.tsx"
+        )
+        val vaadinReactTsx = File(
+            testProject.dir,
+            FrontendUtils.DEFAULT_PROJECT_FRONTEND_GENERATED_DIR + "vaadin-react.tsx"
+        )
+
+        testProject.build("vaadinPrepareFrontend")
+        expect(
+            false,
+            "Expected Flow.tsx not to be created when Hilla is not available"
+        ) { flowTsx.exists() }
+        expect(
+            false,
+            "Expected vaadin-react.tsx not to be created when Hilla is not available"
+        ) { vaadinReactTsx.exists() }
     }
 
     private fun enableHilla() {

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt
@@ -507,10 +507,14 @@ public class PluginEffectiveConfiguration(
 
     public val frontendHotdeploy: Provider<Boolean> =
         extension.frontendHotdeploy
-            .convention(effectiveFrontendDirectory.map { frontendDirectory ->
-                hillaAvailable.get() &&
-                        FrontendUtils.isHillaViewsUsed(frontendDirectory)
-            })
+            .convention(
+                effectiveFrontendDirectory.zip(
+                    hillaAvailable
+                ) { frontendDirectory, hasHilla ->
+                    hasHilla &&
+                            FrontendUtils.isHillaViewsUsed(frontendDirectory)
+                }
+            )
             .overrideWithSystemPropertyFlag(
                 project,
                 InitParameters.FRONTEND_HOTDEPLOY

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt
@@ -19,7 +19,6 @@ import java.io.File
 import java.io.Serializable
 import javax.inject.Inject
 import com.vaadin.flow.internal.StringUtil
-import com.vaadin.flow.plugin.base.BuildFrontendUtil
 import com.vaadin.flow.server.Constants
 import com.vaadin.flow.server.InitParameters
 import com.vaadin.flow.server.frontend.FrontendTools
@@ -31,9 +30,11 @@ import groovy.lang.Closure
 import groovy.lang.DelegatesTo
 import org.gradle.api.Action
 import org.gradle.api.Project
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier
 import org.gradle.api.provider.ListProperty
 import org.gradle.api.provider.Property
 import org.gradle.api.provider.Provider
+import org.gradle.internal.component.external.model.ModuleComponentArtifactIdentifier
 
 public abstract class VaadinFlowPluginExtension @Inject constructor(private val project: Project) {
     /**
@@ -328,7 +329,7 @@ public abstract class VaadinFlowPluginExtension @Inject constructor(private val 
 public class PluginEffectiveConfiguration(
     project: Project,
     extension: VaadinFlowPluginExtension
-) :Serializable {
+) : Serializable {
 
     internal val projectDir = project.projectDir
     internal val projectName = project.name
@@ -339,6 +340,30 @@ public class PluginEffectiveConfiguration(
 
     public val sourceSetName: Property<String> = extension.sourceSetName
         .convention("main")
+
+    public val dependencyScope: Property<String> = extension.dependencyScope
+        .convention(sourceSetName.map {
+            if (it == "main") {
+                "runtimeClasspath"
+            } else {
+                "${it}RuntimeClasspath"
+            }
+        })
+
+
+    internal val hillaAvailable: Provider<Boolean> =
+        project.configurations.getByName(dependencyScope.get())
+            .incoming.artifacts.resolvedArtifacts
+            .map { result ->
+                result.filter {
+                    it.id is ModuleComponentArtifactIdentifier && it.id.componentIdentifier is ModuleComponentIdentifier
+                }.map {
+                    (it.id.componentIdentifier as ModuleComponentIdentifier).moduleIdentifier
+                }.any {
+                    it.group == "com.vaadin" && it.name == "hilla-endpoint"
+                }
+            }
+
 
     public val webpackOutputDirectory: Provider<File> =
         extension.webpackOutputDirectory
@@ -470,15 +495,6 @@ public class PluginEffectiveConfiguration(
 
     public val classpathFilter: ClasspathFilter = extension.classpathFilter
 
-    public val dependencyScope: Property<String> = extension.dependencyScope
-        .convention(sourceSetName.map {
-            if (it == "main") {
-                "runtimeClasspath"
-            } else {
-                "${it}RuntimeClasspath"
-            }
-        })
-
     public val processResourcesTaskName: Property<String> =
         extension.processResourcesTaskName
             .convention(sourceSetName.map {
@@ -491,8 +507,9 @@ public class PluginEffectiveConfiguration(
 
     public val frontendHotdeploy: Provider<Boolean> =
         extension.frontendHotdeploy
-            .convention(effectiveFrontendDirectory.map {
-                FrontendUtils.isHillaUsed(it)
+            .convention(effectiveFrontendDirectory.map { frontendDirectory ->
+                hillaAvailable.get() &&
+                        FrontendUtils.isHillaViewsUsed(frontendDirectory)
             })
             .overrideWithSystemPropertyFlag(
                 project,
@@ -549,7 +566,10 @@ public class PluginEffectiveConfiguration(
 
     public val frontendIgnoreVersionChecks: Provider<Boolean> = extension
         .frontendIgnoreVersionChecks.convention(false)
-        .overrideWithSystemPropertyFlag(project, FrontendUtils.PARAM_IGNORE_VERSION_CHECKS)
+        .overrideWithSystemPropertyFlag(
+            project,
+            FrontendUtils.PARAM_IGNORE_VERSION_CHECKS
+        )
 
     public val npmExcludeWebComponents: Provider<Boolean> = extension
         .npmExcludeWebComponents.convention(false)

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRemoveOldFrontendGeneratedFiles.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRemoveOldFrontendGeneratedFiles.java
@@ -125,12 +125,12 @@ public class TaskRemoveOldFrontendGeneratedFiles implements FallibleCommand {
     }
 
     private Predicate<Path> isKnownUnhandledFile() {
-        Path flowGeneratedImports = FrontendUtils
-                .getFlowGeneratedImports(frontendFolder).toPath()
-                .toAbsolutePath();
-        Path flowGeneratedWebComponentImports = FrontendUtils
+        Path flowGeneratedImports = normalizePath(
+                FrontendUtils.getFlowGeneratedImports(frontendFolder).toPath()
+                        .toAbsolutePath());
+        Path flowGeneratedWebComponentImports = normalizePath(FrontendUtils
                 .getFlowGeneratedWebComponentsImports(frontendFolder).toPath()
-                .toAbsolutePath();
+                .toAbsolutePath());
         Set<Path> knownFiles = new HashSet<>();
         knownFiles.add(flowGeneratedImports);
         knownFiles.add(flowGeneratedWebComponentImports);

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRemoveOldFrontendGeneratedFilesTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRemoveOldFrontendGeneratedFilesTest.java
@@ -151,7 +151,16 @@ public class TaskRemoveOldFrontendGeneratedFilesTest {
 
     @Test
     public void execute_knownFiles_notDeleted() throws Exception {
-        Set<File> knownFiles = Set.of(new File(generatedFolder, "routes.tsx"),
+        Set<File> knownFiles = Set.of(generatedFolder.toPath()
+                .resolve(Path.of("flow", "generated-flow-imports.js")).toFile(),
+                generatedFolder.toPath()
+                        .resolve(Path.of("flow", "generated-flow-imports.d.ts"))
+                        .toFile(),
+                generatedFolder.toPath()
+                        .resolve(Path.of("flow",
+                                "generated-flow-webcomponent-imports.js"))
+                        .toFile(),
+                new File(generatedFolder, "routes.tsx"),
                 new File(generatedFolder, "routes.ts"), generatedFolder.toPath()
                         .resolve(Path.of("flow", "Flow.tsx")).toFile(),
                 new File(generatedFolder, "file-routes.ts"));

